### PR TITLE
[Foundation] Add an improved NSUrlSessionHandlerTrustOverrideCallback that passes the url.

### PR DESCRIPTION
**Pull request by mandel-macaque**
The callback is missing the url wich is useful information to have when
trusting a cert. We do the following:

1. Allow users to use the old callback, but state it as Obsolete.
2. Add a new delegate that will take the url.
3. A cert is accepted if one of the callbacks accepts it. The general
callback is a superset of the url one.

Fixes: https://github.com/xamarin/xamarin-macios/issues/7413

### CODE REVIEW

File: src/Foundation/NSUrlSessionHandler.cs [hunk 1]
```diff
@@ -317,6 +319,17 @@ protected override void Dispose (bool disposing)
 			}
 		}
 
+		NSUrlSessionHandlerTrustOverrideForUrlCallback trustOverrideForUrl;
```
Comment (chamons) [inline src/Foundation/NSUrlSessionHandler.cs]: I don't love the ForUrl suffix, it looks obj-c ish, but I have no better suggestion.

**Comment by mandel-macaque** [comment]
@chamons really hard to write a test for that, not with the current infra

File: src/Foundation/NSUrlSessionHandler.cs [hunk 2]
```diff
@@ -700,8 +713,15 @@ public override void DidReceiveChallenge (NSUrlSession session, NSUrlSessionTask
...
 				var trustCallback = sessionHandler.TrustOverride;
-				if (trustCallback != null && challenge.ProtectionSpace.AuthenticationMethod == NSUrlProtectionSpace.AuthenticationMethodServerTrust) {
-					if (trustCallback (sessionHandler, challenge.ProtectionSpace.ServerSecTrust)) {
+				var trustCallbackForUrl = sessionHandler.TrustOverrideForUrl;
+				var hasCallBack = trustCallback != null || trustCallbackForUrl != null; 
+				if (hasCallBack && challenge.ProtectionSpace.AuthenticationMethod == NSUrlProtectionSpace.AuthenticationMethodServerTrust) {
+					// if one of the delegates allows to ignore the cert, do it. We check first the one that takes the url because is more precisse, later the
+					// more general one. Since we are using nullables, if the delegate is not present, by default is false
+					var trustSec = (trustCallbackForUrl?.Invoke (sessionHandler, inflight.RequestUrl, challenge.ProtectionSpace.ServerSecTrust) ?? false) || 
+						(trustCallback?.Invoke (sessionHandler, challenge.ProtectionSpace.ServerSecTrust) ?? false);
```
Comment (prmallic) [inline src/Foundation/NSUrlSessionHandler.cs]: 👀👍🏽

Comment (whitneyschmidt) [reply src/Foundation/NSUrlSessionHandler.cs]: A+ C# 8 usage ⛳️

**Comment by mandel-macaque** [comment]
Known issue: https://github.com/xamarin/maccore/issues/581

**Comment by nschoenberg** [comment]
@mandel-macaque I use a a root domain certificate i created myself using open ssl, its embedded im my web server that is providing images. I manged to install the cert on my iOS device as custom configuration profile and safari is not showing insecure website anymore.

When I use HTTP Client with NSUrlSession handler on iOS i still get errors that points into direction of tls / insecure certificate. I tryed to use the property TrustOverrideCallback just returning true, but it does not have any effect

**Pull request merged**